### ISSUE TITLE
ParseError on INSTANCE child of ERROR (was: Confusing exception for illegal / unspecified namespaces)

### ISSUE THREAD

**Issue by dugganj** [pywbem/pywbem#1380]
The second call to EnumerateInstances in the attached code generates an exception as it should, since there is no namespace specified -- either in the call directly or a default namespace in the WBEMConnection object.

The problem is that the exception message is horrible.  There is no mention of namespace at all, and the message itself is nonsensical to anyone but a core pywbem developer.

**Comment by dugganj** [comment]
Traceback (most recent call last):
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2018.1.4\helpers\pydev\pydevd.py", line 1664, in <module>
    main()
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2018.1.4\helpers\pydev\pydevd.py", line 1658, in main
    globals = debugger.run(setup['file'], None, None, is_module)
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2018.1.4\helpers\pydev\pydevd.py", line 1068, in run
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2018.1.4\helpers\pydev\_pydev_imps\_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "C:/Users/dugganj/PycharmProjects/untitled/ConfusingNamespaceException.py", line 9, in <module>
    conn.EnumerateInstances("CIM_RegisteredProfile")
  File "C:\Miniconda3\lib\site-packages\pywbem\cim_operations.py", line 2231, in EnumerateInstances
    **extra)
  File "C:\Miniconda3\lib\site-packages\pywbem\cim_operations.py", line 1628, in _imethodcall
    tup_tree = parse_cim(tt_)
  File "C:\Miniconda3\lib\site-packages\pywbem\tupleparse.py", line 336, in parse_cim
    child = one_child(tup_tree, ['MESSAGE', 'DECLARATION'])
  File "C:\Miniconda3\lib\site-packages\pywbem\tupleparse.py", line 229, in one_child
    return parse_any(child)
  File "C:\Miniconda3\lib\site-packages\pywbem\tupleparse.py", line 1919, in parse_any
    return funct_name(tup_tree)
  File "C:\Miniconda3\lib\site-packages\pywbem\tupleparse.py", line 1522, in parse_message
    'MULTIEXPRSP'])
  File "C:\Miniconda3\lib\site-packages\pywbem\tupleparse.py", line 229, in one_child
    return parse_any(child)
  File "C:\Miniconda3\lib\site-packages\pywbem\tupleparse.py", line 1919, in parse_any
    return funct_name(tup_tree)
  File "C:\Miniconda3\lib\site-packages\pywbem\tupleparse.py", line 1758, in parse_simplersp
    child = one_child(tup_tree, ['METHODRESPONSE', 'IMETHODRESPONSE'])
  File "C:\Miniconda3\lib\site-packages\pywbem\tupleparse.py", line 229, in one_child
    return parse_any(child)
  File "C:\Miniconda3\lib\site-packages\pywbem\tupleparse.py", line 1919, in parse_any
    return funct_name(tup_tree)
  File "C:\Miniconda3\lib\site-packages\pywbem\tupleparse.py", line 1811, in parse_imethodresponse
    'PARAMVALUE'])
  File "C:\Miniconda3\lib\site-packages\pywbem\tupleparse.py", line 261, in list_of_various
    result.append(parse_any(child))
  File "C:\Miniconda3\lib\site-packages\pywbem\tupleparse.py", line 1919, in parse_any
    return funct_name(tup_tree)
  File "C:\Miniconda3\lib\site-packages\pywbem\tupleparse.py", line 1825, in parse_error
    check_node(tup_tree, 'ERROR', ['CODE'], ['DESCRIPTION'], [])
  File "C:\Miniconda3\lib\site-packages\pywbem\tupleparse.py", line 195, in check_node
    allowed_children))
pywbem.exceptions.ParseError: Element 'ERROR' has invalid child element 'INSTANCE' (allowed are child elements [])

**Comment by dugganj** [comment]
This is run against the Dell Compellent system in the SMI-Lab at the SNIA tech center.

**Comment by andy-maier** [comment]
@dugganj Jerry, if the `default_namespace` argument is not provided when creating a `WBEMConnection` object, the built-in default "root/cimv2" will be used as a default connection namespace. So when an operation is invoked without specifying a namespace there, that built-in default will be used. Bottom line, the CIM-XML request to the CIM server did have a namespace inside. You can verify that by printing `conn.last_request`.

Therefore, the `ParseError` you get on the response is not caused by a missing namespace on the request. Instead, it is caused by pywbem not supporting the INSTANCE child element under an ERROR element. We need to fix that in pywbem. That is your second bug you found yesterday!

Now that does not yet explain why you got an ERROR elment in your response in the first place, but you could post that here by printing `conn.last_reply`, and then we could have a look at it.

```
try:
    conn.EnumerateInstances("CIM_RegisteredProfile") # this one fails
except pywbem.Error as exc:
    print("Exception: %r" % exc)
    print("last request: %s" % conn.last_request)
    print("last response: %s" % conn.last_reply)
```

**Comment by dugganj** [comment]
Two bugs in one day.  I should get a bug bounty.

I ran with the changes you suggested, and got the spectacularly unhelpful output:

Exception: ParseError("Element 'ERROR' has invalid child element 'INSTANCE' (allowed are child elements [])",)
last request: None
last response: None

**Comment by dugganj** [comment]
# And here is the code...
import pywbem as pw

URL='https://10.1.134.96'
PRINCIPAL='smilab'
PASSWORD='F00sb4ll'

conn = pw.WBEMConnection(URL, (PRINCIPAL, PASSWORD))
conn.EnumerateInstances("CIM_RegisteredProfile", namespace="interop") # this one works
try:
    conn.EnumerateInstances("CIM_RegisteredProfile") # this one fails
except pw.Error as exc:
    print("Exception: %r" % exc)
    print("last request: %s" % conn.last_request)
    print("last response: %s" % conn.last_reply)

**Comment by andy-maier** [comment]
Oops. Forgot to mention that debug needs to be turned on for the connection:
```
conn = pw.WBEMConnection(URL, (PRINCIPAL, PASSWORD))
conn.debug = True
conn.EnumerateInstances("CIM_RegisteredProfile", namespace="interop") # this one works
try:
    conn.EnumerateInstances("CIM_RegisteredProfile") # this one fails
except pw.Error as exc:
print("Exception: %r" % exc)
print("last request: %s" % conn.last_request)
print("last response: %s" % conn.last_reply)
```

**Comment by dugganj** [comment]
Exception: ParseError("Element 'ERROR' has invalid child element 'INSTANCE' (allowed are child elements [])",)
last request: <CIM CIMVERSION="2.0" DTDVERSION="2.0">
  <MESSAGE ID="1001" PROTOCOLVERSION="1.0">
    <SIMPLEREQ>
      <IMETHODCALL NAME="EnumerateInstances">
        <LOCALNAMESPACEPATH>
          <NAMESPACE NAME="root"/>
          <NAMESPACE NAME="cimv2"/>
        </LOCALNAMESPACEPATH>
        <IPARAMVALUE NAME="ClassName">
          <CLASSNAME NAME="CIM_RegisteredProfile"/>
        </IPARAMVALUE>
      </IMETHODCALL>
    </SIMPLEREQ>
  </MESSAGE>
</CIM>

last response: None

**Comment by andy-maier** [comment]
Ok, I still don't know why the last reply is empty, but at least the request shows that the namespace was not missing (LOCALNAMESPACE element). I have a fix for the invalid INSTANCE child element nearly ready.

**Comment by andy-maier** [comment]
PR #1383 fixes the issue with missing support for the INSTANCE child element of the ERROR element, for the upcoming v0.13.0. The fix will be rolled back into 0.12.

**Comment by andy-maier** [comment]
PR #1385 rolls back the fix into 0.12.6. Closing this issue.
